pkg/compose/create.go

```diff
@@ -139,15 +161,42 @@ func (s *composeService) ensureNetworks(ctx context.Context, networks types.Netw
 	return nil
 }
 
-func (s *composeService) ensureProjectVolumes(ctx context.Context, project *types.Project) error {
+func (s *composeService) ensureProjectVolumes(ctx context.Context, project *types.Project, currentState map[string]*volumetypes.Volume, recreateVolumes bool) error {
 	for k, volume := range project.Volumes {
 		volume.Labels = volume.Labels.Add(api.VolumeLabel, k)
 		volume.Labels = volume.Labels.Add(api.ProjectLabel, project.Name)
 		volume.Labels = volume.Labels.Add(api.VersionLabel, api.ComposeVersion)
-		err := s.ensureVolume(ctx, volume, project.Name)
+
+		volumeConfigHash, err := VolumeHash(volume)
+		volume.Labels = volume.Labels.Add(api.VolumeConfigHashLabel, volumeConfigHash)
 		if err != nil {
 			return err
 		}
+		err = s.ensureVolume(ctx, volume, project.Name)
+		if err != nil {
+			return err
+		}
+
+		if recreateVolumes {
+			actualVolume, ok := currentState[k]
+			if !ok {
+				return nil
+			}
+			// Check if it's necessary to recreate the volume by comparing
+			// with the current volume config
+			ok, err = shouldUpdateVolume(volume, actualVolume)
+			if err != nil {
+				return err
+			}
+			if !ok {
+				continue
+			}
+			logrus.Warnf("Warning: Running with --recreate-volumes will delete data for volume %s", k)
```

Comment:
Added a warning message just to emphasize the risk that you will lose the data

Reply:
warning comes too late, as when users see this we already recreated volume.
We should have a warning for user to know volume doesn't match config, and must use `--recreate-volumes` to force compose to recreate accordingly